libsplinter/src/error/constraint_violation.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module containing ConstraintViolationError implementation.
+
+use std::error;
+use std::fmt;
+
+struct Source {
+    source: Box<dyn error::Error>,
+}
+
+/// The type of constraint violation that caused the error
+///
+/// This is the type of constraint on the database's definition that is
+/// violated. For example, if an operation tries to insert an entry that would
+/// cause a duplicate in a unique column, the ConstraintViolationType::Unique
+/// should be used.
+pub enum ConstraintViolationType {
+    Unique,
+    ForeignKey,
+    DataType(Option<String>),
+    Other(String),
+}
+
+impl fmt::Display for ConstraintViolationType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            ConstraintViolationType::Unique => write!(f, "Unique"),
+            ConstraintViolationType::ForeignKey => write!(f, "Foreign Key"),
+            ConstraintViolationType::DataType(ref msg) => match msg {
+                Some(m) => write!(f, "Data Type: {}", m),
+                None => write!(f, "Data Type"),
+            },
+            ConstraintViolationType::Other(ref msg) => write!(f, "Other: {}", msg),
+        }
+    }
+}
+
+impl fmt::Debug for ConstraintViolationType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            ConstraintViolationType::Unique => write!(f, "Unique"),
+            ConstraintViolationType::ForeignKey => write!(f, "Foreign Key"),
+            ConstraintViolationType::DataType(ref msg) => match msg {
+                Some(m) => write!(f, "Data Type: {}", m),
```

Comment:
This will not work well in Debug output, because it will look like 'ConstraintViolationError { violation_type: Data Type: something }, where ':' is separating key/value. Debug should probably be derived instead.